[DTensor] Update redistribute planner cost function based on communication cost

zpcore opened the pull request:
1. Using the same communication cost from `collective_utils.py` when deciding on the min cost redistribution path. 
  - `greedy` and `graph` based redistribution both use `one_step_redistribute_cost` from `collective_utils.py` now. Since the `graph` based redistribution is not enabled by default (unless the device order is nondefault), we won't see any impact.
2. Add the test that resolves #169439.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170109

zpcore committed:
[DTensor] Update redistribute planner cost function based on communication cost

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore commented on test/distributed/tensor/test_redistribute.py:
```diff
@@ -1047,6 +1051,84 @@ def test_ordered_distribute_all_combination(self):
+                # cost function.
+                for i in range(
+                    0, len(shard_orders), 4
+                ):  # we can skip for 2. Skip for 4 to reduce the number of tests
```
I rechecked the test time on A100 host. If I use CUDA, it takes 1m59.918s; If I use localtensor, it takes 2m47.490s. This should be fine. We can use the full test set.

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Support using the communication cost from `collective_utils.py` when deciding on the min cost redistribution path;
2. Allow the context manger to config whether to use the real communication cost or the fixed numeric cost value;
3. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore commented on torch/distributed/tensor/_redistribute.py:
```diff
@@ -283,44 +329,92 @@ def stringify_transform_infos(
-        self.reduce_scatter = reduce_scatter_cost
-        self.chunk_cost = chunk_cost
+        if self.reduce_memory_overhead:
+            all_reduce_cost: float = 4.0
```
I have removed the constant number cost. Now the cost is the same as the number in `collective_utils.py`.

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Using the same communication cost from `collective_utils.py` when deciding on the min cost redistribution path. 
  - `greedy` and `graph` based redistribution both use `one_step_redistribute_cost` from `collective_utils.py` now. Since the `graph` based redistribution is not enabled by default (unless the device order is nondefault), we won't see any impact.
2. Add the test that resolves #169439.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Using the same communication cost from `collective_utils.py` when deciding on the min cost redistribution path. 
  - `greedy` and `graph` based redistribution both use `one_step_redistribute_cost` from `collective_utils.py` now. Since the `graph` based redistribution is not enabled by default (unless the device order is nondefault), we won't see any impact.
2. Add the test that resolves #169439.




[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_redistribute.py:
```diff
@@ -123,14 +127,17 @@ def get_redistribute_planner(
         A DTensorRedistributePlanner instance (potentially cached)
     """
-    cache_key = (weakref.ref(device_mesh), tensor_dimension)
+    if _are_we_tracing():
```
basically this ensures that during tracing we always use non-cached planner right? That's defensive, but will cost us tracing time.  cc @anijain2305 @pianpwk - maybe TAL and see if there is a way to improve so we can use caching while tracing?

zpcore replied:
@pianpwk The error is this:
```
torch._dynamo.exc.InternalTorchDynamoError: RuntimeError: unhashable type: non-nested SymInt
```
Similar to this error: https://www.internalfb.com/phabricator/paste/view/P2017578984

zpcore committed:
Update on "[DTensor] Update redistribute planner cost function based on communication cost"

1. Using the same communication cost from `collective_utils.py` when deciding on the min cost redistribution path. 
  - `greedy` and `graph` based redistribution both use `one_step_redistribute_cost` from `collective_utils.py` now. Since the `graph` based redistribution is not enabled by default (unless the device order is nondefault), we won't see any impact.
2. Add the test that resolves #169439.




[ghstack-poisoned]
